migrations/src/main/scala/za/co/absa/enceladus/migrations/continuous/EntityVersionMap.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2018-2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.migrations.continuous
+
+/**
+  * A base class representing en entity version map. Entity version map is a tool for keeping track
+  * of 'name - version' mapping of entities between different versions of the database.
+  *
+  * It is used for a conflict resolution when doing a continuous migration.
+  * A conflict happens when an entity from the old version of the database has an object id that does not exist
+  * in the new version of the database, but the ‘name - version’ pair of the entity already exists.
+  * Indexes should not allow for ‘name - version’ duplicates to be created.
+  *
+  * * When a conflict happens a new version of the imported entity is created and the mapping between the entity
+  *   in the old database and the new one is created.
+  * * When importing an entity that has a reference to another entity remap the 'foreign key' according to the map.
+  */
+abstract class EntityVersionMap {
+
+  /**
+    * Adds a 'name - version' mapping.
+    *
+    * @param collectionName The name of the collection that containg the entity
+    * @param entityName     An entity name
+    * @param oldVersion     An version of the entity in the old version of the database
+    * @param newVersion     An version of the entity in the new version of the database
+    */
+  @throws[IllegalStateException]
+  final def add(collectionName: String, entityName: String, oldVersion: Int, newVersion: Int): Unit = {
+    get(collectionName, entityName, oldVersion) match {
+      case Some(storedVersion) =>
+        if (storedVersion != newVersion) {
+          throw new IllegalStateException("Attempt to overwrite an entity-version mapping. " +
+          s"collection = '$collectionName', entity = '$entityName', oldVersion = '$oldVersion', " +
+            s"newVersion = '$newVersion'. Version stored = $storedVersion")
```

Comment:
I am not sure I understand from this message which one is the "bad guy"

Reply:
This is going to throw if migration is trying to do something incorrectly. All entity version mappings should be immutable. Once a mapping is created it cannot be changed.

Reply:
👍 Will also add an index to enforce the condition on the database level.

Reply:
Added the index